<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 14 Downloading NCBI sequence data by hand | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 14 Downloading NCBI sequence data by hand | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 14 Downloading NCBI sequence data by hand | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-09-18" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction-to-biological-sequences-databases.html"/>
<link rel="next" href="downloading-sequences-from-uniprot-by-hand.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="downloadR.html"><a href="downloadR.html"><i class="fa fa-check"></i><b>1</b> Downloading R</a>
<ul>
<li class="chapter" data-level="1.1" data-path="downloadR.html"><a href="downloadR.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="downloadR.html"><a href="downloadR.html#introduction-to-r"><i class="fa fa-check"></i><b>1.2</b> Introduction to R</a></li>
<li class="chapter" data-level="1.3" data-path="downloadR.html"><a href="downloadR.html#installing-r"><i class="fa fa-check"></i><b>1.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="downloadR.html"><a href="downloadR.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>1.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="1.3.2" data-path="downloadR.html"><a href="downloadR.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>1.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="downloadR.html"><a href="downloadR.html#starting-r"><i class="fa fa-check"></i><b>1.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html"><i class="fa fa-check"></i><b>2</b> Installing the RStudio IDE {$installRStudio}</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>2.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="2.2" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>2.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="installing-packages.html"><a href="installing-packages.html"><i class="fa fa-check"></i><b>3</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>3.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="3.2" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>3.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="3.3" data-path="installing-packages.html"><a href="installing-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>3.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>4</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>4.1</b> Bioconductor</a></li>
<li class="chapter" data-level="4.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>4.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="4.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>4.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>4.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="4.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>4.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="4.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>4.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>4.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="basic-R.html"><a href="basic-R.html"><i class="fa fa-check"></i><b>5</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="5.1" data-path="basic-R.html"><a href="basic-R.html#vocabulary"><i class="fa fa-check"></i><b>5.1</b> Vocabulary</a></li>
<li class="chapter" data-level="5.2" data-path="basic-R.html"><a href="basic-R.html#r-functions"><i class="fa fa-check"></i><b>5.2</b> R functions</a></li>
<li class="chapter" data-level="5.3" data-path="basic-R.html"><a href="basic-R.html#interacting-with-r"><i class="fa fa-check"></i><b>5.3</b> Interacting with R</a></li>
<li class="chapter" data-level="5.4" data-path="basic-R.html"><a href="basic-R.html#variables-in-r"><i class="fa fa-check"></i><b>5.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="basic-R.html"><a href="basic-R.html#vectors"><i class="fa fa-check"></i><b>5.4.1</b> Vectors</a></li>
<li class="chapter" data-level="5.4.2" data-path="basic-R.html"><a href="basic-R.html#vector-indexing"><i class="fa fa-check"></i><b>5.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="5.4.3" data-path="basic-R.html"><a href="basic-R.html#character-vectors"><i class="fa fa-check"></i><b>5.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="5.4.4" data-path="basic-R.html"><a href="basic-R.html#lists"><i class="fa fa-check"></i><b>5.4.4</b> Lists</a></li>
<li class="chapter" data-level="5.4.5" data-path="basic-R.html"><a href="basic-R.html#tables"><i class="fa fa-check"></i><b>5.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="basic-R.html"><a href="basic-R.html#arguments"><i class="fa fa-check"></i><b>5.5</b> Arguments</a></li>
<li class="chapter" data-level="5.6" data-path="basic-R.html"><a href="basic-R.html#help-files-with-help-and"><i class="fa fa-check"></i><b>5.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="5.7" data-path="basic-R.html"><a href="basic-R.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>5.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="5.8" data-path="basic-R.html"><a href="basic-R.html#more-on-functions"><i class="fa fa-check"></i><b>5.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="5.8.1" data-path="basic-R.html"><a href="basic-R.html#writing-your-own-functions"><i class="fa fa-check"></i><b>5.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="basic-R.html"><a href="basic-R.html#quiting-r"><i class="fa fa-check"></i><b>5.9</b> Quiting R</a></li>
<li class="chapter" data-level="5.10" data-path="basic-R.html"><a href="basic-R.html#links-and-further-reading"><i class="fa fa-check"></i><b>5.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="vectors-in-R.html"><a href="vectors-in-R.html"><i class="fa fa-check"></i><b>6</b> A primer for working with vectors</a>
<ul>
<li class="chapter" data-level="6.1" data-path="vectors-in-R.html"><a href="vectors-in-R.html#preface-1"><i class="fa fa-check"></i><b>6.1</b> Preface</a></li>
<li class="chapter" data-level="6.2" data-path="vectors-in-R.html"><a href="vectors-in-R.html#vocab"><i class="fa fa-check"></i><b>6.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>7</b> Functions</a>
<ul>
<li class="chapter" data-level="7.1" data-path="functions.html"><a href="functions.html#vectors-in-r"><i class="fa fa-check"></i><b>7.1</b> Vectors in R</a></li>
<li class="chapter" data-level="7.2" data-path="functions.html"><a href="functions.html#math-on-vectors"><i class="fa fa-check"></i><b>7.2</b> Math on vectors</a></li>
<li class="chapter" data-level="7.3" data-path="functions.html"><a href="functions.html#functions-on-vectors"><i class="fa fa-check"></i><b>7.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="7.4" data-path="functions.html"><a href="functions.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>7.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="7.5" data-path="functions.html"><a href="functions.html#subsetting-vectors"><i class="fa fa-check"></i><b>7.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="7.6" data-path="functions.html"><a href="functions.html#sequences-of-numbers"><i class="fa fa-check"></i><b>7.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="7.7" data-path="functions.html"><a href="functions.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>7.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="7.8" data-path="functions.html"><a href="functions.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>7.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="plotting-vectors.html"><a href="plotting-vectors.html"><i class="fa fa-check"></i><b>8</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="8.1" data-path="plotting-vectors.html"><a href="plotting-vectors.html#preface-2"><i class="fa fa-check"></i><b>8.1</b> Preface</a></li>
<li class="chapter" data-level="8.2" data-path="plotting-vectors.html"><a href="plotting-vectors.html#plotting-numeric-data"><i class="fa fa-check"></i><b>8.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="8.3" data-path="plotting-vectors.html"><a href="plotting-vectors.html#other-plotting-packages"><i class="fa fa-check"></i><b>8.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="R-objects.html"><a href="R-objects.html"><i class="fa fa-check"></i><b>9</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="9.1" data-path="R-objects.html"><a href="R-objects.html#commands-used"><i class="fa fa-check"></i><b>9.1</b> Commands used</a></li>
<li class="chapter" data-level="9.2" data-path="R-objects.html"><a href="R-objects.html#r-objects"><i class="fa fa-check"></i><b>9.2</b> R Objects</a></li>
<li class="chapter" data-level="9.3" data-path="R-objects.html"><a href="R-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>9.3</b> Differences between objects</a></li>
<li class="chapter" data-level="9.4" data-path="R-objects.html"><a href="R-objects.html#the-data"><i class="fa fa-check"></i><b>9.4</b> The Data</a></li>
<li class="chapter" data-level="9.5" data-path="R-objects.html"><a href="R-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>9.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="R-objects.html"><a href="R-objects.html#is"><i class="fa fa-check"></i><b>9.5.1</b> is()</a></li>
<li class="chapter" data-level="9.5.2" data-path="R-objects.html"><a href="R-objects.html#length"><i class="fa fa-check"></i><b>9.5.2</b> length()</a></li>
<li class="chapter" data-level="9.5.3" data-path="R-objects.html"><a href="R-objects.html#str"><i class="fa fa-check"></i><b>9.5.3</b> str()</a></li>
<li class="chapter" data-level="9.5.4" data-path="R-objects.html"><a href="R-objects.html#c"><i class="fa fa-check"></i><b>9.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="R-objects.html"><a href="R-objects.html#debrief"><i class="fa fa-check"></i><b>9.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html"><i class="fa fa-check"></i><b>10</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="10.1" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-and-rna"><i class="fa fa-check"></i><b>10.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="10.2" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>10.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="10.3" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#rna"><i class="fa fa-check"></i><b>10.3</b> RNA</a></li>
<li class="chapter" data-level="10.4" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#summary"><i class="fa fa-check"></i><b>10.4</b> Summary</a></li>
<li class="chapter" data-level="10.5" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#analysis-questions"><i class="fa fa-check"></i><b>10.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="10.6" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#glossary"><i class="fa fa-check"></i><b>10.6</b> Glossary</a></li>
<li class="chapter" data-level="10.7" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#contributors-and-attributions"><i class="fa fa-check"></i><b>10.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="proteins-review.html"><a href="proteins-review.html"><i class="fa fa-check"></i><b>11</b> Proteins</a>
<ul>
<li class="chapter" data-level="11.0.1" data-path="proteins-review.html"><a href="proteins-review.html#skills-to-develop"><i class="fa fa-check"></i><b>11.0.1</b> Skills to develop:</a></li>
<li class="chapter" data-level="11.1" data-path="proteins-review.html"><a href="proteins-review.html#types-and-functions-of-proteins"><i class="fa fa-check"></i><b>11.1</b> Types and Functions of Proteins</a></li>
<li class="chapter" data-level="11.2" data-path="proteins-review.html"><a href="proteins-review.html#amino-acids"><i class="fa fa-check"></i><b>11.2</b> Amino Acids</a></li>
<li class="chapter" data-level="11.3" data-path="proteins-review.html"><a href="proteins-review.html#evolution-connection"><i class="fa fa-check"></i><b>11.3</b> Evolution Connection:</a></li>
<li class="chapter" data-level="11.4" data-path="proteins-review.html"><a href="proteins-review.html#protein-structure"><i class="fa fa-check"></i><b>11.4</b> Protein Structure</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="proteins-review.html"><a href="proteins-review.html#primary-structure"><i class="fa fa-check"></i><b>11.4.1</b> Primary Structure</a></li>
<li class="chapter" data-level="11.4.2" data-path="proteins-review.html"><a href="proteins-review.html#secondary-structure"><i class="fa fa-check"></i><b>11.4.2</b> Secondary Structure</a></li>
<li class="chapter" data-level="11.4.3" data-path="proteins-review.html"><a href="proteins-review.html#tertiary-structure"><i class="fa fa-check"></i><b>11.4.3</b> Tertiary Structure</a></li>
<li class="chapter" data-level="11.4.4" data-path="proteins-review.html"><a href="proteins-review.html#quaternary-structure"><i class="fa fa-check"></i><b>11.4.4</b> Quaternary Structure</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="proteins-review.html"><a href="proteins-review.html#denaturation-and-protein-folding"><i class="fa fa-check"></i><b>11.5</b> Denaturation and Protein Folding</a></li>
<li class="chapter" data-level="11.6" data-path="proteins-review.html"><a href="proteins-review.html#summary-1"><i class="fa fa-check"></i><b>11.6</b> Summary</a></li>
<li class="chapter" data-level="11.7" data-path="proteins-review.html"><a href="proteins-review.html#art-connections"><i class="fa fa-check"></i><b>11.7</b> Art Connections</a></li>
<li class="chapter" data-level="11.8" data-path="proteins-review.html"><a href="proteins-review.html#review-questions"><i class="fa fa-check"></i><b>11.8</b> Review Questions</a></li>
<li class="chapter" data-level="11.9" data-path="proteins-review.html"><a href="proteins-review.html#free-response"><i class="fa fa-check"></i><b>11.9</b> Free Response</a></li>
<li class="chapter" data-level="11.10" data-path="proteins-review.html"><a href="proteins-review.html#glossary-1"><i class="fa fa-check"></i><b>11.10</b> Glossary</a></li>
<li class="chapter" data-level="11.11" data-path="proteins-review.html"><a href="proteins-review.html#contributors-and-attributions-1"><i class="fa fa-check"></i><b>11.11</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html"><i class="fa fa-check"></i><b>12</b> NCBI: The National Center for Biotechnology Information</a>
<ul>
<li class="chapter" data-level="12.1" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#genbank-sequence-database"><i class="fa fa-check"></i><b>12.1</b> GenBank sequence database</a></li>
<li class="chapter" data-level="12.2" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#pubmed-and-pubmed-central-article-database"><i class="fa fa-check"></i><b>12.2</b> PubMed and PubMed Central article database</a></li>
<li class="chapter" data-level="12.3" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#entrez"><i class="fa fa-check"></i><b>12.3</b> Entrez</a></li>
<li class="chapter" data-level="12.4" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#blast"><i class="fa fa-check"></i><b>12.4</b> BLAST</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html"><i class="fa fa-check"></i><b>13</b> Introduction to biological sequences databases</a>
<ul>
<li class="chapter" data-level="13.1" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#topics"><i class="fa fa-check"></i><b>13.1</b> Topics</a></li>
<li class="chapter" data-level="13.2" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#introduction"><i class="fa fa-check"></i><b>13.2</b> Introduction</a></li>
<li class="chapter" data-level="13.3" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#biological-sequence-databases"><i class="fa fa-check"></i><b>13.3</b> Biological sequence databases</a></li>
<li class="chapter" data-level="13.4" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>13.4</b> The NCBI Sequence Database</a></li>
<li class="chapter" data-level="13.5" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sub-databases"><i class="fa fa-check"></i><b>13.5</b> The NCBI Sub-Databases</a></li>
<li class="chapter" data-level="13.6" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#ncbi-genbank-record-format"><i class="fa fa-check"></i><b>13.6</b> NCBI GenBank Record Format</a></li>
<li class="chapter" data-level="13.7" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-fasta-file-format"><i class="fa fa-check"></i><b>13.7</b> The FASTA file format</a></li>
<li class="chapter" data-level="13.8" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#refseq"><i class="fa fa-check"></i><b>13.8</b> RefSeq</a></li>
<li class="chapter" data-level="13.9" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#querying-the-ncbi-database"><i class="fa fa-check"></i><b>13.9</b> Querying the NCBI Database</a></li>
<li class="chapter" data-level="13.10" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#querying-the-ncbi-database-via-the-ncbi-website-for-reference"><i class="fa fa-check"></i><b>13.10</b> Querying the NCBI Database via the NCBI Website (for reference)</a></li>
<li class="chapter" data-level="13.11" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#example-finding-the-sequences-published-in-nature-460352-358-for-reference"><i class="fa fa-check"></i><b>13.11</b> Example: finding the sequences published in Nature 460:352-358 (for reference)</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html"><i class="fa fa-check"></i><b>14</b> Downloading NCBI sequence data by hand</a>
<ul>
<li class="chapter" data-level="14.1" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html#preface-3"><i class="fa fa-check"></i><b>14.1</b> Preface</a></li>
<li class="chapter" data-level="14.2" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html#retrieving-genome-sequence-data-via-the-ncbi-website"><i class="fa fa-check"></i><b>14.2</b> Retrieving genome sequence data via the NCBI website</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html"><i class="fa fa-check"></i><b>15</b> Downloading sequences from UniProt by hand</a>
<ul>
<li class="chapter" data-level="15.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#vocab-1"><i class="fa fa-check"></i><b>15.1</b> Vocab</a></li>
<li class="chapter" data-level="15.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#downloading-protein-data-from-uniprot"><i class="fa fa-check"></i><b>15.2</b> Downloading Protein data from UniProt</a></li>
<li class="chapter" data-level="15.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#viewing-the-uniprot-webpage-for-a-protein-sequence"><i class="fa fa-check"></i><b>15.3</b> Viewing the UniProt webpage for a protein sequence</a>
<ul>
<li class="chapter" data-level="15.3.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-function"><i class="fa fa-check"></i><b>15.3.1</b> Protein function</a></li>
<li class="chapter" data-level="15.3.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-sequence-and-size"><i class="fa fa-check"></i><b>15.3.2</b> Protein sequence and size</a></li>
<li class="chapter" data-level="15.3.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#other-information"><i class="fa fa-check"></i><b>15.3.3</b> Other information</a></li>
</ul></li>
<li class="chapter" data-level="15.4" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#retrieving-a-uniprot-protein-sequence-via-the-uniprot-website"><i class="fa fa-check"></i><b>15.4</b> Retrieving a UniProt protein sequence via the UniProt website</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="introducingFASTA.html"><a href="introducingFASTA.html"><i class="fa fa-check"></i><b>16</b> Introducing FASTA Files</a>
<ul>
<li class="chapter" data-level="16.1" data-path="introducingFASTA.html"><a href="introducingFASTA.html#example-fasta-file"><i class="fa fa-check"></i><b>16.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="16.2" data-path="introducingFASTA.html"><a href="introducingFASTA.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>16.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="16.3" data-path="introducingFASTA.html"><a href="introducingFASTA.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>16.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="16.4" data-path="introducingFASTA.html"><a href="introducingFASTA.html#fastq-format"><i class="fa fa-check"></i><b>16.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>17</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="18" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>18</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="18.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-1"><i class="fa fa-check"></i><b>18.1</b> Introduction</a></li>
<li class="chapter" data-level="18.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>18.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-2"><i class="fa fa-check"></i><b>18.2.1</b> Vocab</a></li>
<li class="chapter" data-level="18.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>18.2.2</b> R functions</a></li>
<li class="chapter" data-level="18.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>18.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="18.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>18.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>18.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="18.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>18.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="18.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>18.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="18.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>18.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="18.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>18.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="18.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>18.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="18.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>18.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="18.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>18.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="18.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>18.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="18.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>18.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="18.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>18.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>19</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>19.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="19.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>19.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="19.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>19.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="19.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>19.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="19.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>19.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-1"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="19.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>19.2</b> Help!</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>19.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="19.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>19.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="19.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>19.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="19.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>19.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>19.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="19.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>19.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="19.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>19.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>19.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="downloading-ncbi-sequence-data-by-hand" class="section level1" number="14">
<h1><span class="header-section-number">Chapter 14</span> Downloading NCBI sequence data by hand</h1>
<p><strong>By</strong>: Avril Coghlan.</p>
<p><strong>Adapted, edited and expanded</strong>: Nathan Brouwer under the Creative Commons 3.0 Attribution License <a href="https://creativecommons.org/licenses/by/3.0/">(CC BY 3.0)</a>.</p>
<div id="preface-3" class="section level2" number="14.1">
<h2><span class="header-section-number">14.1</span> Preface</h2>
<p>The following chapter was originally written by Avril Coghlan. It provides brief, basic information for how to access sequences via the internet. Subsequent chapters provide more details and R code.</p>
</div>
<div id="retrieving-genome-sequence-data-via-the-ncbi-website" class="section level2" number="14.2">
<h2><span class="header-section-number">14.2</span> Retrieving genome sequence data via the NCBI website</h2>
<p>You can easily retrieve DNA or protein sequence data by hand from the <a href="www.ncbi.nlm.nih.gov">NCBI</a> Sequence Database via its website www.ncbi.nlm.nih.gov.</p>
<p>Dengue DEN-1 DNA is a viral DNA sequence and its NCBI <strong>accession number</strong> is NC_001477. To retrieve the DNA sequence for the Dengue DEN-1 virus from NCBI, go to the NCBI website, type “NC_001477” in the Search box at the top of the webpage, and press the “Search” button beside the Search box.</p>
<p>(While this is the normal workflow, accessions related to well-known organisms can sometimes turn up using a direct Google search. This is the case for NC_001477 - if you Google it you can go directly to the website with the genome sequence.)</p>
<p>On the results page of a normal NCBI search you will see the number of hits to “NC_001477” in each of the NCBI databases on the NCBI website. There are <em>many</em> databases on the NCBI website, for example, <strong>PubMed</strong> and <strong>Pubmed Central</strong> contain abstracts from scientific papers, the <strong>Genes</strong> and <strong>Genomes</strong> database contains DNA and RNA sequence data, the <strong>Proteins</strong> database contains protein sequence data, and so on.</p>
<p>Most biologist would do this type of work by hand from within their web browser, but it can also be done by writing small programs in <strong>scripting languages</strong> such as <strong>Python</strong> or <strong>R</strong>. In R, the <strong>rentrez</strong> package is a powerful tool for intersecting with NCBI resource. In this tutorial we’ll focus on the web interface. Its good to remember, though, that almost anything done via the webpage can be automated using a computer <strong>script</strong>.</p>
<p>A challenge when learning to use NCBI resources is that there is a tremendous amount of sequence information available and you need to learn how to sort through what the search results provide. As you are looking for the <em>DNA</em> sequence of the Dengue DEN-1 virus genome, you expect to see a hit in the NCBI <strong>Nucleotide database</strong>. This is indicated at the top of the page where it says “<strong>NUCLEOTIDE SEQUENCE</strong>” and lists “Dengue virus 1, complete genome.”</p>
<p>When you click on the link for the Nucleotide database, it will bring you to the record for NC_001477 in the NCBI Nucleotide database. This will contain the name and NCBI <strong>accession</strong> of the sequence, as well as other details such as any papers describing the sequence. If you scroll down you’ll see the sequence also.</p>
<p>If you need it, you can retrieve the DNA sequence for the DEN-1 Dengue virus genome sequence as a <strong>FASTA format</strong> sequence file in a couple ways. The easiest is just to copy and paste it into a text, .R, or other file. You can also click on <strong>“Send to”</strong> at the top right of the NC_001477 sequence record webpage (just to the left of the side bar; its kinda small).</p>
<p>After you click on <strong>Send to</strong> you can pick several options. and then choose “File” in the menu that appears, and then choose FASTA from the “Format” menu that appears, and click on “Create file”. The sequence will then download. The default file name is <strong>sequence.fasta</strong> so you’ll probably want to change it.</p>
<p>You can now open the FASTA file containing the DEN-1 Dengue virus genome sequence using a text editor like Notepad, WordPad, Notepad++, or even RStudio on your computer. To find a text editor on your computer search for “text” from the start menu (Windows) and usually one will come up. (Opening the file in a word processor like word isn’t recommended).</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-biological-sequences-databases.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="downloading-sequences-from-uniprot-by-hand.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/004-NCBI/03-NCBI_seqdata_by_GUI1.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/004-NCBI/03-NCBI_seqdata_by_GUI1.Rmd",
"text": null
},
"download": ["lbrb.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
