<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Nucleic Acids | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Nucleic Acids | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Nucleic Acids | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avirl Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2021-09-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="basic-R.html"/>
<link rel="next" href="proteins-review.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="downloadR.html"><a href="downloadR.html"><i class="fa fa-check"></i><b>1</b> Downloading R</a>
<ul>
<li class="chapter" data-level="1.1" data-path="downloadR.html"><a href="downloadR.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="downloadR.html"><a href="downloadR.html#introduction-to-r"><i class="fa fa-check"></i><b>1.2</b> Introduction to R</a></li>
<li class="chapter" data-level="1.3" data-path="downloadR.html"><a href="downloadR.html#installing-r"><i class="fa fa-check"></i><b>1.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="downloadR.html"><a href="downloadR.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>1.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="1.3.2" data-path="downloadR.html"><a href="downloadR.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>1.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="downloadR.html"><a href="downloadR.html#starting-r"><i class="fa fa-check"></i><b>1.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html"><i class="fa fa-check"></i><b>2</b> Installing the RStudio IDE {$installRStudio}</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>2.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="2.2" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>2.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="installing-packages.html"><a href="installing-packages.html"><i class="fa fa-check"></i><b>3</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>3.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="3.2" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>3.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="3.3" data-path="installing-packages.html"><a href="installing-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>3.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>4</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>4.1</b> Bioconductor</a></li>
<li class="chapter" data-level="4.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>4.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="4.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>4.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>4.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="4.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>4.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="4.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>4.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>4.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="basic-R.html"><a href="basic-R.html"><i class="fa fa-check"></i><b>5</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="5.1" data-path="basic-R.html"><a href="basic-R.html#vocabulary"><i class="fa fa-check"></i><b>5.1</b> Vocabulary</a></li>
<li class="chapter" data-level="5.2" data-path="basic-R.html"><a href="basic-R.html#r-functions"><i class="fa fa-check"></i><b>5.2</b> R functions</a></li>
<li class="chapter" data-level="5.3" data-path="basic-R.html"><a href="basic-R.html#interacting-with-r"><i class="fa fa-check"></i><b>5.3</b> Interacting with R</a></li>
<li class="chapter" data-level="5.4" data-path="basic-R.html"><a href="basic-R.html#variables-in-r"><i class="fa fa-check"></i><b>5.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="basic-R.html"><a href="basic-R.html#vectors"><i class="fa fa-check"></i><b>5.4.1</b> Vectors</a></li>
<li class="chapter" data-level="5.4.2" data-path="basic-R.html"><a href="basic-R.html#vector-indexing"><i class="fa fa-check"></i><b>5.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="5.4.3" data-path="basic-R.html"><a href="basic-R.html#character-vectors"><i class="fa fa-check"></i><b>5.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="5.4.4" data-path="basic-R.html"><a href="basic-R.html#lists"><i class="fa fa-check"></i><b>5.4.4</b> Lists</a></li>
<li class="chapter" data-level="5.4.5" data-path="basic-R.html"><a href="basic-R.html#tables"><i class="fa fa-check"></i><b>5.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="basic-R.html"><a href="basic-R.html#arguments"><i class="fa fa-check"></i><b>5.5</b> Arguments</a></li>
<li class="chapter" data-level="5.6" data-path="basic-R.html"><a href="basic-R.html#help-files-with-help-and"><i class="fa fa-check"></i><b>5.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="5.7" data-path="basic-R.html"><a href="basic-R.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>5.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="5.8" data-path="basic-R.html"><a href="basic-R.html#more-on-functions"><i class="fa fa-check"></i><b>5.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="5.8.1" data-path="basic-R.html"><a href="basic-R.html#writing-your-own-functions"><i class="fa fa-check"></i><b>5.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="basic-R.html"><a href="basic-R.html#quiting-r"><i class="fa fa-check"></i><b>5.9</b> Quiting R</a></li>
<li class="chapter" data-level="5.10" data-path="basic-R.html"><a href="basic-R.html#links-and-further-reading"><i class="fa fa-check"></i><b>5.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html"><i class="fa fa-check"></i><b>6</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="6.1" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-and-rna"><i class="fa fa-check"></i><b>6.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="6.2" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>6.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="6.3" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#rna"><i class="fa fa-check"></i><b>6.3</b> RNA</a></li>
<li class="chapter" data-level="6.4" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#summary"><i class="fa fa-check"></i><b>6.4</b> Summary</a></li>
<li class="chapter" data-level="6.5" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#analysis-questions"><i class="fa fa-check"></i><b>6.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="6.6" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#glossary"><i class="fa fa-check"></i><b>6.6</b> Glossary</a></li>
<li class="chapter" data-level="6.7" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#contributors-and-attributions"><i class="fa fa-check"></i><b>6.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="proteins-review.html"><a href="proteins-review.html"><i class="fa fa-check"></i><b>7</b> Proteins</a>
<ul>
<li class="chapter" data-level="7.0.1" data-path="proteins-review.html"><a href="proteins-review.html#skills-to-develop"><i class="fa fa-check"></i><b>7.0.1</b> Skills to develop:</a></li>
<li class="chapter" data-level="7.1" data-path="proteins-review.html"><a href="proteins-review.html#types-and-functions-of-proteins"><i class="fa fa-check"></i><b>7.1</b> Types and Functions of Proteins</a></li>
<li class="chapter" data-level="7.2" data-path="proteins-review.html"><a href="proteins-review.html#amino-acids"><i class="fa fa-check"></i><b>7.2</b> Amino Acids</a></li>
<li class="chapter" data-level="7.3" data-path="proteins-review.html"><a href="proteins-review.html#evolution-connection"><i class="fa fa-check"></i><b>7.3</b> Evolution Connection:</a></li>
<li class="chapter" data-level="7.4" data-path="proteins-review.html"><a href="proteins-review.html#protein-structure"><i class="fa fa-check"></i><b>7.4</b> Protein Structure</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="proteins-review.html"><a href="proteins-review.html#primary-structure"><i class="fa fa-check"></i><b>7.4.1</b> Primary Structure</a></li>
<li class="chapter" data-level="7.4.2" data-path="proteins-review.html"><a href="proteins-review.html#secondary-structure"><i class="fa fa-check"></i><b>7.4.2</b> Secondary Structure</a></li>
<li class="chapter" data-level="7.4.3" data-path="proteins-review.html"><a href="proteins-review.html#tertiary-structure"><i class="fa fa-check"></i><b>7.4.3</b> Tertiary Structure</a></li>
<li class="chapter" data-level="7.4.4" data-path="proteins-review.html"><a href="proteins-review.html#quaternary-structure"><i class="fa fa-check"></i><b>7.4.4</b> Quaternary Structure</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="proteins-review.html"><a href="proteins-review.html#denaturation-and-protein-folding"><i class="fa fa-check"></i><b>7.5</b> Denaturation and Protein Folding</a></li>
<li class="chapter" data-level="7.6" data-path="proteins-review.html"><a href="proteins-review.html#summary-1"><i class="fa fa-check"></i><b>7.6</b> Summary</a></li>
<li class="chapter" data-level="7.7" data-path="proteins-review.html"><a href="proteins-review.html#art-connections"><i class="fa fa-check"></i><b>7.7</b> Art Connections</a></li>
<li class="chapter" data-level="7.8" data-path="proteins-review.html"><a href="proteins-review.html#review-questions"><i class="fa fa-check"></i><b>7.8</b> Review Questions</a></li>
<li class="chapter" data-level="7.9" data-path="proteins-review.html"><a href="proteins-review.html#free-response"><i class="fa fa-check"></i><b>7.9</b> Free Response</a></li>
<li class="chapter" data-level="7.10" data-path="proteins-review.html"><a href="proteins-review.html#glossary-1"><i class="fa fa-check"></i><b>7.10</b> Glossary</a></li>
<li class="chapter" data-level="7.11" data-path="proteins-review.html"><a href="proteins-review.html#contributors-and-attributions-1"><i class="fa fa-check"></i><b>7.11</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectors-in-R.html"><a href="vectors-in-R.html"><i class="fa fa-check"></i><b>8</b> A primer for working with vectors</a>
<ul>
<li class="chapter" data-level="8.1" data-path="vectors-in-R.html"><a href="vectors-in-R.html#preface-1"><i class="fa fa-check"></i><b>8.1</b> Preface</a></li>
<li class="chapter" data-level="8.2" data-path="vectors-in-R.html"><a href="vectors-in-R.html#vocab"><i class="fa fa-check"></i><b>8.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>9</b> Functions</a>
<ul>
<li class="chapter" data-level="9.1" data-path="functions.html"><a href="functions.html#vectors-in-r"><i class="fa fa-check"></i><b>9.1</b> Vectors in R</a></li>
<li class="chapter" data-level="9.2" data-path="functions.html"><a href="functions.html#math-on-vectors"><i class="fa fa-check"></i><b>9.2</b> Math on vectors</a></li>
<li class="chapter" data-level="9.3" data-path="functions.html"><a href="functions.html#functions-on-vectors"><i class="fa fa-check"></i><b>9.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="9.4" data-path="functions.html"><a href="functions.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>9.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="9.5" data-path="functions.html"><a href="functions.html#subsetting-vectors"><i class="fa fa-check"></i><b>9.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="9.6" data-path="functions.html"><a href="functions.html#sequences-of-numbers"><i class="fa fa-check"></i><b>9.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="9.7" data-path="functions.html"><a href="functions.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>9.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="9.8" data-path="functions.html"><a href="functions.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>9.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="plotting-vectors.html"><a href="plotting-vectors.html"><i class="fa fa-check"></i><b>10</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="10.1" data-path="plotting-vectors.html"><a href="plotting-vectors.html#preface-2"><i class="fa fa-check"></i><b>10.1</b> Preface</a></li>
<li class="chapter" data-level="10.2" data-path="plotting-vectors.html"><a href="plotting-vectors.html#plotting-numeric-data"><i class="fa fa-check"></i><b>10.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="10.3" data-path="plotting-vectors.html"><a href="plotting-vectors.html#other-plotting-packages"><i class="fa fa-check"></i><b>10.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="R-objects.html"><a href="R-objects.html"><i class="fa fa-check"></i><b>11</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="11.1" data-path="R-objects.html"><a href="R-objects.html#commands-used"><i class="fa fa-check"></i><b>11.1</b> Commands used</a></li>
<li class="chapter" data-level="11.2" data-path="R-objects.html"><a href="R-objects.html#r-objects"><i class="fa fa-check"></i><b>11.2</b> R Objects</a></li>
<li class="chapter" data-level="11.3" data-path="R-objects.html"><a href="R-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>11.3</b> Differences between objects</a></li>
<li class="chapter" data-level="11.4" data-path="R-objects.html"><a href="R-objects.html#the-data"><i class="fa fa-check"></i><b>11.4</b> The Data</a></li>
<li class="chapter" data-level="11.5" data-path="R-objects.html"><a href="R-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>11.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="R-objects.html"><a href="R-objects.html#is"><i class="fa fa-check"></i><b>11.5.1</b> is()</a></li>
<li class="chapter" data-level="11.5.2" data-path="R-objects.html"><a href="R-objects.html#length"><i class="fa fa-check"></i><b>11.5.2</b> length()</a></li>
<li class="chapter" data-level="11.5.3" data-path="R-objects.html"><a href="R-objects.html#str"><i class="fa fa-check"></i><b>11.5.3</b> str()</a></li>
<li class="chapter" data-level="11.5.4" data-path="R-objects.html"><a href="R-objects.html#c"><i class="fa fa-check"></i><b>11.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="R-objects.html"><a href="R-objects.html#debrief"><i class="fa fa-check"></i><b>11.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html"><i class="fa fa-check"></i><b>12</b> NCBI: The National Center for Biotechnology Information</a>
<ul>
<li class="chapter" data-level="12.1" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#genbank-sequence-database"><i class="fa fa-check"></i><b>12.1</b> GenBank sequence database</a></li>
<li class="chapter" data-level="12.2" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#pubmed-and-pubmed-central-article-database"><i class="fa fa-check"></i><b>12.2</b> PubMed and PubMed Central article database</a></li>
<li class="chapter" data-level="12.3" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#entrez"><i class="fa fa-check"></i><b>12.3</b> Entrez</a></li>
<li class="chapter" data-level="12.4" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#blast"><i class="fa fa-check"></i><b>12.4</b> BLAST</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html"><i class="fa fa-check"></i><b>13</b> Introduction to biological sequences databases</a>
<ul>
<li class="chapter" data-level="13.1" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#topics"><i class="fa fa-check"></i><b>13.1</b> Topics</a></li>
<li class="chapter" data-level="13.2" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#introduction"><i class="fa fa-check"></i><b>13.2</b> Introduction</a></li>
<li class="chapter" data-level="13.3" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#biological-sequence-databases"><i class="fa fa-check"></i><b>13.3</b> Biological sequence databases</a></li>
<li class="chapter" data-level="13.4" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>13.4</b> The NCBI Sequence Database</a></li>
<li class="chapter" data-level="13.5" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sub-databases"><i class="fa fa-check"></i><b>13.5</b> The NCBI Sub-Databases</a></li>
<li class="chapter" data-level="13.6" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#ncbi-genbank-record-format"><i class="fa fa-check"></i><b>13.6</b> NCBI GenBank Record Format</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html"><a href="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html"><i class="fa fa-check"></i><b>14</b> <code>{r, echo = F, eval = F} # knitr::include_graphics(here::here("images/NCBI_accesssion_NC_001477_genbank.png")) #</code></a>
<ul>
<li class="chapter" data-level="14.1" data-path="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html"><a href="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html#the-fasta-file-format"><i class="fa fa-check"></i><b>14.1</b> The FASTA file format</a></li>
<li class="chapter" data-level="14.2" data-path="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html"><a href="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html#refseq"><i class="fa fa-check"></i><b>14.2</b> RefSeq</a></li>
<li class="chapter" data-level="14.3" data-path="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html"><a href="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html#querying-the-ncbi-database"><i class="fa fa-check"></i><b>14.3</b> Querying the NCBI Database</a></li>
<li class="chapter" data-level="14.4" data-path="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html"><a href="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html#querying-the-ncbi-database-via-the-ncbi-website-for-reference"><i class="fa fa-check"></i><b>14.4</b> Querying the NCBI Database via the NCBI Website (for reference)</a></li>
<li class="chapter" data-level="14.5" data-path="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html"><a href="r-echo-f-eval-f-knitrinclude-graphicsherehereimagesncbi-accesssion-nc-001477-genbank-png.html#example-finding-the-sequences-published-in-nature-460352-358-for-reference"><i class="fa fa-check"></i><b>14.5</b> Example: finding the sequences published in Nature 460:352-358 (for reference)</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html"><i class="fa fa-check"></i><b>15</b> Downloading NCBI sequence data by hand</a>
<ul>
<li class="chapter" data-level="15.1" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html#preface-3"><i class="fa fa-check"></i><b>15.1</b> Preface</a></li>
<li class="chapter" data-level="15.2" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html#retrieving-genome-sequence-data-via-the-ncbi-website"><i class="fa fa-check"></i><b>15.2</b> Retrieving genome sequence data via the NCBI website</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html"><i class="fa fa-check"></i><b>16</b> Downloading sequences from UniProt by hand</a>
<ul>
<li class="chapter" data-level="16.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#vocab-1"><i class="fa fa-check"></i><b>16.1</b> Vocab</a></li>
<li class="chapter" data-level="16.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#downloading-protein-data-from-uniprot"><i class="fa fa-check"></i><b>16.2</b> Downloading Protein data from UniProt</a></li>
<li class="chapter" data-level="16.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#viewing-the-uniprot-webpage-for-a-protein-sequence"><i class="fa fa-check"></i><b>16.3</b> Viewing the UniProt webpage for a protein sequence</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-function"><i class="fa fa-check"></i><b>16.3.1</b> Protein function</a></li>
<li class="chapter" data-level="16.3.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-sequence-and-size"><i class="fa fa-check"></i><b>16.3.2</b> Protein sequence and size</a></li>
<li class="chapter" data-level="16.3.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#other-information"><i class="fa fa-check"></i><b>16.3.3</b> Other information</a></li>
</ul></li>
<li class="chapter" data-level="16.4" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#retrieving-a-uniprot-protein-sequence-via-the-uniprot-website"><i class="fa fa-check"></i><b>16.4</b> Retrieving a UniProt protein sequence via the UniProt website</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html"><i class="fa fa-check"></i><b>17</b> Introducing FASTA Files</a>
<ul>
<li class="chapter" data-level="17.1" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#example-fasta-file"><i class="fa fa-check"></i><b>17.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="17.2" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>17.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="17.3" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>17.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="17.4" data-path="introducing-FASTA.html"><a href="introducing-FASTA.html#fastq-format"><i class="fa fa-check"></i><b>17.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html"><i class="fa fa-check"></i><b>18</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="18.0.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#functions-1"><i class="fa fa-check"></i><b>18.0.1</b> Functions</a></li>
<li class="chapter" data-level="18.0.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#softwarewebsites"><i class="fa fa-check"></i><b>18.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="18.0.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#r-vocabulary"><i class="fa fa-check"></i><b>18.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="18.0.4" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#file-types"><i class="fa fa-check"></i><b>18.0.4</b> File types</a></li>
<li class="chapter" data-level="18.0.5" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>18.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="18.0.6" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>18.0.6</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="18.0.7" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#preliminaries"><i class="fa fa-check"></i><b>18.0.7</b> Preliminaries</a></li>
<li class="chapter" data-level="18.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>18.1</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>18.1.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="18.1.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>18.1.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="18.1.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#fasta-file-format"><i class="fa fa-check"></i><b>18.1.3</b> FASTA file format</a></li>
<li class="chapter" data-level="18.1.4" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#the-ncbi-sequence-database-1"><i class="fa fa-check"></i><b>18.1.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="18.1.5" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>18.1.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#optional-saving-fasta-files"><i class="fa fa-check"></i><b>18.2</b> OPTIONAL: Saving FASTA files</a></li>
<li class="chapter" data-level="18.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#next-steps"><i class="fa fa-check"></i><b>18.3</b> Next steps</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html"><i class="fa fa-check"></i><b>19</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="19.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>19.1</b> Preliminaries</a></li>
<li class="chapter" data-level="19.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>19.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="19.2.1" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>19.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="19.2.2" data-path="downloading-dna-sequences-as-fasta-files-in-r.html"><a href="downloading-dna-sequences-as-fasta-files-in-r.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>19.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html"><i class="fa fa-check"></i><b>20</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="20.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preface-4"><i class="fa fa-check"></i><b>20.1</b> Preface</a></li>
<li class="chapter" data-level="20.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#writing-todo"><i class="fa fa-check"></i><b>20.2</b> Writing TODO:</a></li>
<li class="chapter" data-level="20.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#introduction-1"><i class="fa fa-check"></i><b>20.3</b> Introduction</a></li>
<li class="chapter" data-level="20.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>20.4</b> Vocabulary</a></li>
<li class="chapter" data-level="20.5" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#functions-2"><i class="fa fa-check"></i><b>20.5</b> Functions</a></li>
<li class="chapter" data-level="20.6" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#preliminaries-2"><i class="fa fa-check"></i><b>20.6</b> Preliminaries</a></li>
<li class="chapter" data-level="20.7" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>20.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="20.8" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>20.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="20.8.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>20.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="20.8.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>20.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="20.8.3" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#dna-words"><i class="fa fa-check"></i><b>20.8.3</b> DNA words</a></li>
<li class="chapter" data-level="20.8.4" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#summary-2"><i class="fa fa-check"></i><b>20.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="20.9" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#acknowledgements"><i class="fa fa-check"></i><b>20.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="20.9.1" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#license"><i class="fa fa-check"></i><b>20.9.1</b> License</a></li>
<li class="chapter" data-level="20.9.2" data-path="dna-descriptive-statics-part-1.html"><a href="dna-descriptive-statics-part-1.html#exercises"><i class="fa fa-check"></i><b>20.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="download-protein-seqs-R.html"><a href="download-protein-seqs-R.html"><i class="fa fa-check"></i><b>21</b> Downloading protein sequences in <em>R</em></a>
<ul>
<li class="chapter" data-level="21.1" data-path="download-protein-seqs-R.html"><a href="download-protein-seqs-R.html#preliminaries-3"><i class="fa fa-check"></i><b>21.1</b> Preliminaries</a></li>
<li class="chapter" data-level="21.2" data-path="download-protein-seqs-R.html"><a href="download-protein-seqs-R.html#retrieving-a-uniprot-protein-sequence-using-rentrez"><i class="fa fa-check"></i><b>21.2</b> Retrieving a UniProt protein sequence using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="changes-to-database-entries.html"><a href="changes-to-database-entries.html"><i class="fa fa-check"></i><b>22</b> Changes to database entries</a></li>
<li class="chapter" data-level="23" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>23</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="24" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>24</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="24.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-2"><i class="fa fa-check"></i><b>24.1</b> Introduction</a></li>
<li class="chapter" data-level="24.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>24.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="24.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-2"><i class="fa fa-check"></i><b>24.2.1</b> Vocab</a></li>
<li class="chapter" data-level="24.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>24.2.2</b> R functions</a></li>
<li class="chapter" data-level="24.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>24.2.3</b> Download necessary packages</a></li>
<li class="chapter" data-level="24.2.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>24.2.4</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="24.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>24.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="24.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>24.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="24.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>24.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="24.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>24.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="24.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>24.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="24.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>24.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="24.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>24.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="24.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>24.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="24.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distance."><i class="fa fa-check"></i><b>24.9</b> Genetic distance.</a></li>
<li class="chapter" data-level="24.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-finally"><i class="fa fa-check"></i><b>24.10</b> Phylognetic trees (finally!)</a>
<ul>
<li class="chapter" data-level="24.10.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>24.10.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="25" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>25</b> Logarithms in R</a></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 01: Getting access to R</a>
<ul>
<li class="chapter" data-level="25.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>25.1</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="25.1.1" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>25.1.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="25.1.2" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>25.1.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="25.1.3" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>25.1.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="25.1.4" data-path="appendix-01-getting-access-to-r.html"><a href="appendix-01-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>25.1.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="25.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>25.2</b> Help!</a>
<ul>
<li class="chapter" data-level="25.2.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>25.2.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="25.2.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>25.2.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="25.2.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>25.2.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="25.2.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>25.2.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="25.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>25.3</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="25.3.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>25.3.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="25.3.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>25.3.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="25.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>25.4</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="nucleic-acids-review" class="section level1" number="6">
<h1><span class="header-section-number">Chapter 6</span> Nucleic Acids</h1>
<p><strong>Authors</strong>: OpenStax / Libretext Formatted in RMarkdown by Nathan Brouwer under the Creative Commons Attribution License 4.0 license.</p>
<p>This chapter was adapted from <a href="https://bio.libretexts.org/Bookshelves/Introductory_and_General_Biology/Book%3A_General_Biology_(OpenStax)">LibreText General Biology</a>, Chapter 3, Section 3.5: <a href="https://bio.libretexts.org/Bookshelves/Introductory_and_General_Biology/Book%3A_General_Biology_(OpenStax)/1%3A_The_Chemistry_of_Life/3%3A_Biological_Macromolecules/3.5%3A_Nucleic_Acids">Nucleic Acids</a>. The LibreText book is based on <a href="https://openstax.org/details/books/biology-2e">OpenStax Biology 2nd edition</a>, Chapter 3, Section 3.5: <a href="https://openstax.org/books/biology-2e/pages/3-5-nucleic-acids">Nucleic Acids</a>. A full list of authors is found under the <strong>Contributors and Attributions</strong> section at the end of this document.</p>
<blockquote>
<p>Nucleic acids are the most important macromolecules for the continuity
of life. They carry the genetic blueprint of a cell and carry
instructions for the functioning of the cell.</p>
</blockquote>
<div id="dna-and-rna" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> DNA and RNA</h2>
<p>The two main types of nucleic acids are deoxyribonucleic acid (DNA)
and ribonucleic acid (RNA). DNA is the genetic material found in all
living organisms, ranging from single-celled bacteria to multicellular
mammals. It is found in the nucleus of eukaryotes and in the
organelles, chloroplasts, and mitochondria. In prokaryotes, the DNA is
not enclosed in a membranous envelope.</p>
<p>The entire genetic content of a cell is known as its genome, and the
study of genomes is genomics. In eukaryotic cells but not in
prokaryotes, DNA forms a complex with histone proteins to form
chromatin, the substance of eukaryotic chromosomes. A chromosome may
contain tens of thousands of genes. Many genes contain the information
to make protein products; other genes code for RNA products. DNA
controls all of the cellular activities by turning the genes “on” or
“off.”</p>
<p>The other type of nucleic acid, RNA, is mostly involved in protein
synthesis. The DNA molecules never leave the nucleus but instead use
an intermediary to communicate with the rest of the cell. This
intermediary is the messenger RNA (mRNA). Other types of RNA—like
rRNA, tRNA, and microRNA—are involved in protein synthesis and its
regulation.</p>
<p>DNA and RNA are made up of monomers known as nucleotides. The
nucleotides combine with each other to form a polynucleotide, DNA or
RNA. Each nucleotide is made up of three components: a nitrogenous
base, a pentose (five-carbon) sugar, and a phosphate group (Figure
3.5.1). Each nitrogenous base in a nucleotide is attached to a sugar
molecule, which is attached to one or more phosphate groups.</p>
<div class="figure"><span id="fig:unnamed-chunk-49"></span>
<img src="images/image5.jpeg" alt="A nucleotide is made up of three components: a nitrogenous base, a pentose sugar, and one or more phosphate groups. Carbon residues in the pentose are numbered 1′ through 5′ (the prime distinguishes these residues from those in the base, which are numbered without using a prime notation). The base is attached to the 1′ position of the ribose, and the phosphate is attached to the 5′ position. When a polynucleotide is formed, the 5′ phosphate of the incoming nucleotide attaches to the 3′ hydroxyl group at the end of the growing chain. Two types of pentose are found in nucleotides, deoxyribose (found in DNA) and ribose (found in RNA). Deoxyribose is similar in structure to ribose, but it has an H instead of an OH at the 2′ position. Bases can be divided into two categories: purines and pyrimidines. Purines have a double ring structure, and pyrimidines have a single ring." width="536" />
<p class="caption">
Figure 6.1: A nucleotide is made up of three components: a nitrogenous base, a pentose sugar, and one or more phosphate groups. Carbon residues in the pentose are numbered 1′ through 5′ (the prime distinguishes these residues from those in the base, which are numbered without using a prime notation). The base is attached to the 1′ position of the ribose, and the phosphate is attached to the 5′ position. When a polynucleotide is formed, the 5′ phosphate of the incoming nucleotide attaches to the 3′ hydroxyl group at the end of the growing chain. Two types of pentose are found in nucleotides, deoxyribose (found in DNA) and ribose (found in RNA). Deoxyribose is similar in structure to ribose, but it has an H instead of an OH at the 2′ position. Bases can be divided into two categories: purines and pyrimidines. Purines have a double ring structure, and pyrimidines have a single ring.
</p>
</div>
<p>The nitrogenous bases, important components of nucleotides, are
organic molecules and are so named because they contain carbon and
nitrogen. They are bases because they contain an amino group that has
the potential of binding an extra hydrogen, and thus, decreases the
hydrogen ion concentration in its environment, making it more basic.
Each nucleotide in DNA contains one of four possible nitrogenous
bases: adenine (A), guanine (G) cytosine (C), and thymine (T).</p>
<p>Adenine and guanine are classified as purines. The primary structure
of a purine is two carbon-nitrogen rings. Cytosine, thymine, and
uracil are classified as pyrimidines which have a single
carbon-nitrogen ring as their primary structure (Figure 3.5.1). Each
of these basic carbon-nitrogen rings has different functional groups
attached to it. In molecular biology shorthand, the nitrogenous bases are simply known by their symbols A, T, G, C, and U. DNA contains A, T, G, and C whereas RNA contains A, U, G, and
C.</p>
<p>The pentose sugar in DNA is deoxyribose, and in RNA, the sugar is
ribose (Figure 3.5.1). The difference between the sugars is the
presence of the hydroxyl group on the second carbon of the ribose and
hydrogen on the second carbon of the deoxyribose. The carbon atoms of
the sugar molecule are numbered as 1′, 2′, 3′, 4′, and 5′ (1′ is read
as “one prime”). The phosphate residue is attached to the hydroxyl
group of the 5′ carbon of one sugar and the hydroxyl group of the 3′
carbon of the sugar of the next nucleotide, which forms a 5′–3′
phosphodiester linkage. The phosphodiester linkage is not formed by
simple dehydration reaction like the other linkages connecting
monomers in macromolecules: its formation involves the removal of two
phosphate groups. A polynucleotide may have thousands of such
phosphodiester linkages.</p>
</div>
<div id="dna-double-helix-structure" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> DNA Double-Helix Structure</h2>
<p>DNA has a double-helix structure (Figure 3.5.2). The sugar and
phosphate lie on the outside of the helix, forming the backbone of the
DNA. The nitrogenous bases are stacked in the interior, like the steps
of a staircase, in pairs; the pairs are bound to each other by
hydrogen bonds. Every base pair in the double helix is separated from
the next base pair by 0.34 nm. The two strands of the helix run in
opposite directions, meaning that the 5′ carbon end of one strand will
face the 3′ carbon end of its matching strand. (This is referred to as
antiparallel orientation and is important to DNA replication and in
many nucleic acid interactions.)</p>
<div class="figure"><span id="fig:unnamed-chunk-50"></span>
<img src="images/image6.jpeg" alt="Figure 3.5.2 : Native DNA is an antiparallel double helix. The phosphate backbone (indicated by the curvy lines) is on the outside, and the bases are on the inside. Each base from one strand interacts via hydrogen bonding with a base from the opposing strand. (credit: Jerome Walker/Dennis Myts)" width="272" />
<p class="caption">
Figure 6.2: Figure 3.5.2 : Native DNA is an antiparallel double helix. The phosphate backbone (indicated by the curvy lines) is on the outside, and the bases are on the inside. Each base from one strand interacts via hydrogen bonding with a base from the opposing strand. (credit: Jerome Walker/Dennis Myts)
</p>
</div>
<p>Only certain types of base pairing are allowed. For example, a certain
purine can only pair with a certain pyrimidine. This means A can pair
with T, and G can pair with C, as shown in Figure 3.5.3. This is known
as the base complementary rule. In other words, the DNA strands are
complementary to each other. If the sequence of one strand is
AATTGGCC, the complementary strand would have the sequence TTAACCGG.
During DNA replication, each strand is copied, resulting in a daughter
DNA double helix containing one parental DNA strand and a newly
synthesized strand.</p>
<p>Art Connection
<img src="images/image7.jpeg" width="234" /></p>
<p>Figure 3.5.3 : In a double stranded DNA molecule, the two strands run
antiparallel to one another so that one strand runs 5′ to 3′ and the
other 3′ to 5′. The phosphate backbone is located on the outside, and
the bases are in the middle. Adenine forms hydrogen bonds (or base
pairs) with thymine, and guanine base pairs with cytosine.</p>
<p>A mutation occurs, and cytosine is replaced with adenine. What impact
do you think this will have on the DNA structure?</p>
</div>
<div id="rna" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> RNA</h2>
<p>Ribonucleic acid, or RNA, is mainly involved in the process of protein
synthesis under the direction of DNA. RNA is usually single-stranded
and is made of ribonucleotides that are linked by phosphodiester
bonds. A ribonucleotide in the RNA chain contains ribose (the pentose
sugar), one of the four nitrogenous bases (A, U, G, and C), and the
phosphate group.</p>
<p>There are four major types of RNA: messenger RNA (mRNA), ribosomal RNA
(rRNA), transfer RNA (tRNA), and microRNA (miRNA). The first, mRNA,
carries the message from DNA, which controls all of the cellular
activities in a cell. If a cell requires a certain protein to be
synthesized, the gene for this product is turned “on” and the
messenger RNA is synthesized in the nucleus. The RNA base sequence is
complementary to the coding sequence of the DNA from which it has been
copied. However, in RNA, the base T is absent and U is present
instead. If the DNA strand has a sequence AATTGCGC, the sequence of
the complementary RNA is UUAACGCG. In the cytoplasm, the mRNA
interacts with ribosomes and other cellular machinery (Figure 3.5.4).</p>
<p><img src="images/image8.jpeg" width="256" /></p>
<p>Figure 3.5.4 : A ribosome has two parts: a large subunit and a small
subunit. The mRNA sits in between the two subunits. A tRNA molecule
recognizes a codon on the mRNA, binds to it by complementary base
pairing, and adds the correct amino acid to the growing peptide chain.</p>
<p>The mRNA is read in sets of three bases known as codons. Each codon
codes for a single amino acid. In this way, the mRNA is read and the
protein product is made. Ribosomal RNA (rRNA) is a major constituent
of ribosomes on which the mRNA binds. The rRNA ensures the proper
alignment of the mRNA and the ribosomes; the rRNA of the ribosome also
has an enzymatic activity (peptidyl transferase) and catalyzes the
formation of the peptide bonds between two aligned amino acids.
Transfer RNA (tRNA) is one of the smallest of the four types of RNA,
usually 70–90 nucleotides long. It carries the correct amino acid to
the site of protein synthesis. It is the base pairing between the tRNA
and mRNA that allows for the correct amino acid to be inserted in the
polypeptide chain. microRNAs are the smallest RNA molecules and their
role involves the regulation of gene expression by interfering with
the expression of certain mRNA messages. Table 3.5.1 below summarizes
features of DNA and RNA.</p>
<p><strong><em>Table</em></strong> 3.5.1 *<strong>:</strong> Features of DNA and RNA.</p>
<table>
<colgroup>
<col width="30%" />
<col width="34%" />
<col width="34%" />
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Features of DNA and
RNA</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td></td>
<td>DNA</td>
<td>RNA</td>
</tr>
<tr class="even">
<td>Function</td>
<td>Carries genetic
information</td>
<td>Involved in protein
synthesis</td>
</tr>
<tr class="odd">
<td>Location</td>
<td>Remains in the nucleus</td>
<td>Leaves the nucleus</td>
</tr>
<tr class="even">
<td>Structure</td>
<td>Double helix</td>
<td>Usually
single-stranded</td>
</tr>
<tr class="odd">
<td>Sugar</td>
<td>Deoxyribose</td>
<td>Ribose</td>
</tr>
<tr class="even">
<td>Pyrimidines</td>
<td>Cytosine, thymine</td>
<td>Cytosine, uracil</td>
</tr>
<tr class="odd">
<td>Purines</td>
<td>Adenine, guanine</td>
<td>Adenine, guanine</td>
</tr>
</tbody>
</table>
<p>Even though the RNA is single stranded, most RNA types show extensive
intramolecular base pairing between complementary sequences, creating
a predictable three-dimensional structure essential for their
function.</p>
<p>As you have learned, information flow in an organism takes place from
DNA to RNA to protein. DNA dictates the structure of mRNA in a process
known as transcription, and RNA dictates the structure of protein in a
process known as translation. This is known as the Central Dogma of
Life, which holds true for all organisms; however, exceptions to the
rule occur in connection with viral infections.</p>
<p><em>Link to Learning</em>
<img src="images/image9.png" width="10%" /></p>
<ul>
<li>To learn more about DNA, explore the Howard Hughes Medical Institute
BioInteractive animations on the topic of DNA.*</li>
</ul>
</div>
<div id="summary" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Summary</h2>
<p>Nucleic acids are molecules made up of nucleotides that direct
cellular activities such as cell division and protein synthesis. Each
nucleotide is made up of a pentose sugar, a nitrogenous base, and a
phosphate group. There are two types of nucleic acids: DNA and RNA.
DNA carries the genetic blueprint of the cell and is passed on from
parents to offspring (in the form of chromosomes). It has a
double-helical structure with the two strands running in opposite
directions, connected by hydrogen bonds, and complementary to each
other. RNA is single-stranded and is made of a pentose sugar (ribose),
a nitrogenous base, and a phosphate group. RNA is involved in protein
synthesis and its regulation. Messenger RNA (mRNA) is copied from the
DNA, is exported from the nucleus to the cytoplasm, and contains
information for the construction of proteins. Ribosomal RNA (rRNA) is
a part of the ribosomes at the site of protein synthesis, whereas
transfer RNA (tRNA) carries the amino acid to the site of protein
synthesis. microRNA regulates the use of mRNA for protein synthesis.</p>
<p>Art Connections</p>
</div>
<div id="analysis-questions" class="section level2" number="6.5">
<h2><span class="header-section-number">6.5</span> Analysis Questions</h2>
<p>Free Response</p>
</div>
<div id="glossary" class="section level2" number="6.6">
<h2><span class="header-section-number">6.6</span> Glossary</h2>
<p><strong>deoxyribonucleic acid (DNA)</strong>: double-helical molecule that carries the hereditary information of the cell<br />
<strong>messenger RNA (mRNA)</strong>: DNA that carries information from DNA to ribosomes during protein synthesis<br />
<strong>nucleic acid</strong>: biological macromolecule that carries the genetic blueprint of a cell
and carries instructions for the functioning of the cell<br />
<strong>nucleotide</strong>: monomer of nucleic acids; contains a pentose sugar, one or more phosphate groups, and a nitrogenous base<br />
<strong>phosphodiester</strong>: linkage covalent chemical bond that holds together the polynucleotide
chains with a phosphate group linking two pentose sugars of neighboring nucleotides<br />
<strong>polynucleotide</strong>: long chain of nucleotides<br />
<strong>purine</strong>: type of nitrogenous base in DNA and RNA; adenine and guanine are<br />
<strong>purines, pyrimidine</strong>: type of nitrogenous base in DNA and RNA; cytosine, thymine, and uracil
are pyrimidines<br />
<strong>ribonucleic acid (RNA)</strong>: single-stranded, often internally base paired, molecule that is
involved in protein synthesis<br />
<strong>ribosomal RNA (rRNA)</strong>: RNA that ensures the proper alignment of the mRNA and the ribosomes
during protein synthesis and catalyzes the formation of the peptide linkage<br />
<strong>transcription</strong>: process through which messenger RNA forms on a template of DNA<br />
<strong>transfer RNA (tRNA)</strong>: RNA that carries activated amino acids to the site of protein
synthesis on the ribosome<br />
<strong>translation</strong>: process through which RNA directs the formation of protein</p>
</div>
<div id="contributors-and-attributions" class="section level2" number="6.7">
<h2><span class="header-section-number">6.7</span> Contributors and Attributions</h2>
<p>Connie Rye (East Mississippi Community College), Robert Wise
(University of Wisconsin, Oshkosh), Vladimir Jurukovski (Suffolk
County Community College), Jean DeSaix (University of North Carolina
at Chapel Hill), Jung Choi (Georgia Institute of Technology), Yael
Avissar (Rhode Island College) among other contributing authors.
Original content by OpenStax (CC BY 4.0; Download for free at
<a href="http://cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd%409.87">http://cnx.org/contents/185cbf87-c72...f21b5eabd@9.87</a>).</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="basic-R.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="proteins-review.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/002-biology_review/00g-review-nucleic_acids-LibreText.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/002-biology_review/00g-review-nucleic_acids-LibreText.Rmd",
"text": null
},
"download": ["lbrb.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
